code/parachain/frame/composable-maths/src/dex/tests.rs

```diff
@@ -0,0 +1,162 @@
+				a_sent: 117,
+				fee: 12,
+			},
+			InputsAndOutputs {
+				w_i: Permill::from_percent(34),
+				w_o: Permill::from_percent(66),
+				b_i: 1024,
+				b_o: 2048,
+				a_out: 100,
+				f: Permill::from_percent(10),
+				a_sent: 117,
+				fee: 12,
+			},
+			InputsAndOutputs {
+				w_i: Permill::from_percent(25),
+				w_o: Permill::from_percent(75),
+				b_i: 10_000,
+				b_o: 30_000,
+				a_out: 100,
+				f: Permill::from_percent(10),
+				a_sent: 112,
+				fee: 12,
+			},
+			InputsAndOutputs {
+				w_i: Permill::from_percent(40),
+				w_o: Permill::from_percent(60),
+				b_i: 20_000_000,
+				b_o: 30_000_000,
+				a_out: 100_000,
+				f: Permill::from_percent(10),
+				a_sent: 111_576,
+				fee: 11_158,
+			},
+			InputsAndOutputs {
+				w_i: Permill::from_percent(40),
+				w_o: Permill::from_percent(20),
+				b_i: 20_000_000,
+				b_o: 10_000_000,
+				a_out: 100_000,
+				f: Permill::from_percent(10),
+				a_sent: 111_952,
+				fee: 11_196,
+			},
+		];
+
+		prop_compose! {
+			fn checked_inputs_and_outputs()
+			(x in 0..CHECKED_I_AND_O_LIST.len()) -> InputsAndOutputs {
+				CHECKED_I_AND_O_LIST[x]
+			}
+		}
+
+		#[test]
+		fn should_return_zero_fee_when_f_is_zero() {
+			let w_i = Permill::from_rational::<u32>(1, 2);
+			let w_o = Permill::from_rational::<u32>(1, 2);
+			let b_i = 12;
+			let b_o = 12;
+			let a_out = 2;
+			let f = Permill::zero();
+
+			let res = compute_in_given_out_new(w_i, w_o, b_i, b_o, a_out, f)
+				.expect("Input is valid; QED");
+
+			assert_eq!(res.1, 0);
+		}
+
+		#[test]
+		fn should_return_error_when_w_i_is_zero() {
+			let w_i = Permill::zero();
+			let w_o = Permill::from_rational::<u32>(1, 2);
+			let b_i = 12;
+			let b_o = 12;
+			let a_out = 2;
+			let f = Permill::from_percent(10);
+
+			let res = compute_in_given_out_new(w_i, w_o, b_i, b_o, a_out, f);
+
+			assert_eq!(res, Err(InGivenOutError::from(ArithmeticError::DivisionByZero)));
+		}
+
+		#[test]
+		fn should_return_error_when_b_o_and_a_out_are_zero() {
+			let w_i = Permill::from_rational::<u32>(1, 2);
+			let w_o = Permill::from_rational::<u32>(1, 2);
+			let b_i = 12;
+			let b_o = 0;
+			let a_out = 0;
+			let f = Permill::from_percent(10);
+
+			let res = compute_in_given_out_new(w_i, w_o, b_i, b_o, a_out, f);
+
+			assert_eq!(res, Err(InGivenOutError::from(ArithmeticError::DivisionByZero)));
+		}
+
+		#[test]
+		fn should_return_error_when_a_out_is_greater_than_b_o() {
+			let w_i = Permill::from_rational::<u32>(1, 2);
+			let w_o = Permill::from_rational::<u32>(1, 2);
+			let b_i = 512;
+			let b_o = 128;
+			let a_out = 256;
+			let f = Permill::from_percent(10);
+
+			let res = compute_in_given_out_new(w_i, w_o, b_i, b_o, a_out, f);
+
+			assert_eq!(res, Err(InGivenOutError::CanNotTakeMoreThanAvailable))
+		}
+
+		proptest! {
+			#![proptest_config(ProptestConfig::with_cases(CHECKED_I_AND_O_LIST.len() as u32))]
+
+			#[test]
+			fn should_pass_with_expected_values(i_and_o in checked_inputs_and_outputs()) {
```

Comment:
Same NIT comment as the other PR, we could have number ranges tested for unexpected errors in combinations

Reply:
86412d5845e06dc9dac38a4f447de303492dc8ad